src/app/withdraw/components/util/validate-withdraw-amount.ts

```diff
@@ -0,0 +1,107 @@
+import getEmilyLimits from "@/actions/get-emily-limits";
+import yup from "yup";
+type EmilyLimits = Awaited<ReturnType<typeof getEmilyLimits>>;
+
+const getMaxError = (maxWithdrawal: number) => {
+  return `Withdrawal exceeds current cap of ${maxWithdrawal.toLocaleString(undefined, { maximumFractionDigits: 8 })} BTC`;
+};
+function convertBTCtoSats(btcValue: number) {
+  return btcValue * 1e8;
+}
+
+function checkWithdrawalCap(
+  this: yup.TestContext<yup.AnyObject>,
+  satsValue: number,
+  emilyLimits: EmilyLimits | undefined,
+) {
+  if (
+    emilyLimits?.perWithdrawalCap &&
+    satsValue > emilyLimits.perWithdrawalCap
+  ) {
+    return this.createError({
+      path: this.path,
+      message: getMaxError(emilyLimits.perWithdrawalCap / 1e8),
+    });
+  }
+  return true;
+}
+
+function checkAvailableToWithdraw(
+  this: yup.TestContext<yup.AnyObject>,
+  satsValue: number,
+  emilyLimits: EmilyLimits | undefined,
+) {
+  if (
+    emilyLimits?.availableToWithdraw &&
+    satsValue > emilyLimits.availableToWithdraw
+  ) {
+    return this.createError({
+      path: this.path,
+      message: getMaxError(emilyLimits.availableToWithdraw / 1e8),
+    });
+  }
+  return true;
+}
+
+function checkWithdrawableBalance(
+  this: yup.TestContext<yup.AnyObject>,
+  btcValue: number,
+  btcBalance: number,
+  fee: number,
+) {
+  const maxWithdrawableBalance = btcBalance - fee;
+  if (btcValue > maxWithdrawableBalance) {
+    return this.createError({
+      path: this.path,
+      message: formatBalanceError(btcValue, fee, btcBalance),
+    });
+  }
+  return true;
+}
+
+function formatBalanceError(btcValue: number, fee: number, btcBalance: number) {
+  return `The withdrawal + max fees (${(fee + btcValue).toLocaleString(
+    undefined,
+    {
+      maximumFractionDigits: 8,
+    },
+  )}) amount exceeds your current balance of ${btcBalance.toLocaleString(
+    undefined,
+    {
+      maximumFractionDigits: 8,
+    },
+  )} sBTC`;
+}
+
+// Main validation function
+export function validateWithdrawal(
+  this: yup.TestContext<yup.AnyObject>,
+  btcValue: number | undefined,
+  emilyLimits: EmilyLimits | undefined,
```

Comment:
should we bail if `emilyLimits` is undefined, instead of skipping the checks?

Reply:
good idea, since we must have emily up for withdrawals to work

Reply:
fixed